src/librustc_trans/back/link.rs

```diff
@@ -237,7 +237,7 @@ pub fn default_output_for_target(sess: &Session) -> config::CrateType {
 /// Checks if target supports crate_type as output
 pub fn invalid_output_for_target(sess: &Session,
                                  crate_type: config::CrateType) -> bool {
-    match (sess.target.target.options.dynamic_linking,
+    match (sess.target.target.options.dynamic_linking && !sess.target_is_crt_static(),
```

Comment:
Hm I think this isn't quite accurate for MSVC where you can statically link msvcrt but still create DLLs (I believe)